clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java

```
@@ -790,6 +812,197 @@ public void testOnSubscriptionUpdatedTransitionsToJoiningOnlyIfNotInGroup() {
         verify(membershipManager, never()).transitionToJoining();
     }
 
+    @Test
+    public void testListenerCallbacksBasic() {
```

Should we add the following tests?
* Calling on partitions assigned with partitions.
* Receiving an error when on revoked is called.
* Receiving an error when on assigned is called.

---

Will do.

---

@dajac @lianetm—I'm seeing an issue when writing a test that validates failures for `onPartitionsRevoked()`. If an `onPartitionsRevoked()` callback throws an exception, is the intention that we still run the `onPartitionsAssigned()` callback regardless?

---

Yes, that's the intention to keep the behaviour of the legacy coordinator, where `onPartitionsAssigned` is always invoked, even if the `onPartitionsRevoked` fails.

---

Hum... @lianetm In this case, it seems that the reconciliation is not failed when only `onPartitionsRevoked` fails. Is this right? It may be better to just fail the reconciliation when `onPartitionsRevoked` fails. What do you think?

---

Agree @dajac , reconciliation should fail if the `onPartitionsRevoked` fails, we're indeed missing that path here (it's only happening when `onPartitionsAssigned` fails). 

When `onPartitionsRevoked` fails, we intentionally invoke the `onPartitionsAssigned` anyways as I was mentioning above (experience the legacy consumer provides), but we should end up failing the reconciliation as it happens when `onPartitionsAssign` fails. Makes sense @kirktrue ?

---

I wonder if we should just fail the reconciliation process when `onPartitionsRevoked` fails. What do you think, @lianetm?

---

@lianet said:

> When `onPartitionsRevoked` fails, we intentionally invoke the `onPartitionsAssigned` anyways as I was mentioning above (experience the legacy consumer provides), but we should end up failing the reconciliation as it happens when `onPartitionsAssign` fails. Makes sense @kirktrue ?

This isn't what I'm seeing. In my new `MembershipManagerTest` `testListenerCallbacksThrowsErrorOnPartitionsRevoked`, if an exception is thrown from `onPartitionsRevoked()`, the reconciliation step for assignment is skipped completely such that `onPartitionsAssigned()` is not called. Is that a bug or no?

---

@kirktrue the current implementation ignores the `onPartitionsRevoked` callback result [here](https://github.com/apache/kafka/blob/bc4e2aaf3381297fbe2e947acac3ed692c7e7da0/clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java#L752) and continues to `assignPartitions` to invoke `onPartitionsAssigned`, so looks like a bug (`onPartitionsRevoked` completion failing differently from what was expected?)

That being said, @dajac is bringing out a valid point. We were intentionally continuing in this case of failure, just to call both callbacks as the legacy logic does, but I don't see any clear value in it. Just failing the reconciliation when `onPartitionsRevoked` fails seems more consistent to me: the client failed to reconcile (revoke), so it does not carry on with the reconciliation, and does not ack to the broker. Stays RECONCILING, with the `assignmentReadyToReconcile`, and we should ensure that it is all processed again on the next reconciliation loop.

If we agree on that, it means that we should handle the revocation result [here](https://github.com/apache/kafka/blob/bc4e2aaf3381297fbe2e947acac3ed692c7e7da0/clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java#L752) where it is currently ignored, and just do what it is already done when the `onPartitionsAssgined` fails [here](https://github.com/apache/kafka/blob/bc4e2aaf3381297fbe2e947acac3ed692c7e7da0/clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java#L777). Makes sense @kirktrue ?

---

Actually I was missing the fact that the [thenCompose](https://github.com/apache/kafka/blob/bc4e2aaf3381297fbe2e947acac3ed692c7e7da0/clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java#L752) after the revocation is probably the one short-circuiting, so we do get the behaviour we want, of just aborting the reconciliation. So we can leave it like it is and enable the test. 

As a follow-up I would make the logic clearer, by having both callbacks leading to the failed reconciliation in the same way, but we can leave that out of this PR. I can handle that when I unify the callbacks with [KAFKA-15835](https://issues.apache.org/jira/browse/KAFKA-15835)

---

Yes, the 'assigned' block that lives inside `thenCompose()` is never called if the parent `Future` was completed exceptionally.

If I'm reading this correctly, we want to change the existing behavior for the `AsyncKafkaConsumer` to _not_ call `onPartitionsAssigned()` if `onPartitionsRevoked()` throws an exception. I'll update my tests accordingly.

That said, I imagine one or more of the integration tests will fail with this change, so they will need to remain enabled only for `GENERIC`.

---

If we cannot reuse the existing integration tests, we will need to add new ones. We could do it in a separate PR though.

---

I've updated the `MembershipManagerImplTest` to include these latest changes. Thanks for clarifying the desired behavior.